### PR TITLE
Experimental/template based generator

### PR DESCRIPTION
**Pull request by teone**
When working at adding support to the nested list I realized that I was going down a path that would have caused me to basically replicate all the information present in `yang.Entry` in a different format. Which makes no sense.

I'm taking a stab at generating the code passing `yang.Entry` directly to the template, and it seems pretty smooth for `leaves` and `containers`. I'll tackle the `lists` Monday, and hopefully it won't take forever.

Meanwhile opening the PR in case @SeanCondon notices anything completely wrong with this approach.

There is also a stab at unit-testing the generated code, as working of the `testdevice` or other models becomes complicated pretty quickly and might not cover corner cases.

### CODE REVIEW

**Comment by SeanCondon** [comment]
> Meanwhile opening the PR in case @SeanCondon notices anything completely wrong with this approach. 

I have to say I'm finding it increasingly difficult to follow this approach. I think it needs a diagram at this stage with templates referencing templates referencing packages.

If you find you are recreating the dirEntry tree of yang.Entry, you could do what I did on the XPath work with a 2 step approach - which is to take a run through the tree as a first pass, and add any information you might want to the `Annotation` map at the end of each Entry - then on a second pass you might be able to navigate through it in a simpler way.

Overall though I think what you're trying to achieve here is to generate Go code that will simplify the mapping of the API model to YANG model in GraphQL and gRPC - and some of that you will always have to do by hand because the models could be very different. So maybe the payback on extra automation here is limited, and could be handled in the hand-coding stage instead. Just a suggestion, if you're getting config fatigue.

**Comment by ray-milkey** [comment]
retest this please

**Comment by teone** [comment]
@SeanCondon I finally got the nested lists to work fine for the `aether-models`. As of now only the `Get` methods are implemented, but I'd like to merge before starting to work on the `Delete` and `Update` methods.

I added some documentation in the `pkg/gnmi-client-gen/README.md` file and I'd be happy to discuss more on this approach in case things are not clear yet.

Let me know what you think

**Comment by SeanCondon** [comment]
> Let me know what you think

@teone the documentation is very helpful in making it clearer - thanks for adding it
It looks like there are just a few copyright headers needed and it's good

File: pkg/gnmi-client-gen/generator_test.go [hunk 1]
```diff
@@ -7,202 +7,261 @@
 package gnmi_client_gen
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
+	testdevice "github.com/onosproject/config-models/models/testdevice-1.0.x/api"
```
Comment (SeanCondon) [inline pkg/gnmi-client-gen/generator_test.go:13]: this creates a looped dependency that could prevent further updates of YGOT

You could just import the []byte from generated.go like I did in `config-models/pkg/path/extract_test.go`

**Review by SeanCondon — APPROVED** [review]
One comment about a test - approved

File: pkg/gnmi-client-gen/generator_test.go (hunk 1 above)
Comment (teone) [reply pkg/gnmi-client-gen/generator_test.go:13]: I can remove this test altogether, I just needed it to extract the schema as it was helpful to see the entire structure while designing the templates, but it's not required and/or actually testing anything.

Just curious, why would it prevent updates to YGOT?

Comment (SeanCondon) [reply pkg/gnmi-client-gen/generator_test.go:13]: in the last upgrade the schema []byte was changed by the YGOT upgrade. So if that was to happen again the testdevice, would have to be upgraded before config-models could be updated, and vice versa, creating a deadlock situation

Comment (teone) [reply pkg/gnmi-client-gen/generator_test.go:13]: I think they could be upgraded at the same time as they are in the same package, but let me remove that tests as it's not actually a test anyway

**Pull request merged**